File: backend/src/sellpilot/schemas/selection.py
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import Field, model_validator

from sellpilot.core.enums import AnalysisStatus, CurrencyCode, SiteCode
from sellpilot.schemas.common import ContractModel

SelectionSortField = Literal["sales_count", "rating", "review_count", "price", "updated_at"]


class SelectionCandidateQuery(ContractModel):
    site: SiteCode
    category_id: str | None = Field(default=None, min_length=1, max_length=100)
    product_ids: list[str] | None = Field(default=None, min_length=1, max_length=100)
    min_price: Decimal | None = Field(default=None, ge=0)
    max_price: Decimal | None = Field(default=None, gt=0)
    sort_by: SelectionSortField = "sales_count"
    descending: bool = True
    offset: int = Field(default=0, ge=0)
    limit: int = Field(default=100, ge=1, le=100)

    @model_validator(mode="after")
    def validate_price_range(self) -> "SelectionCandidateQuery":
        if self.min_price is not None and self.max_price is not None:
            if self.min_price > self.max_price:
                raise ValueError("min_price must not exceed max_price")
        return self


class SelectionAnalysisRequest(SelectionCandidateQuery):
    minimum_profit: Decimal = Decimal("0")
    minimum_margin: Decimal = Field(default=Decimal("0"), ge=-1, le=1)
    platform_fee_rate: Decimal = Field(default=Decimal("0"), ge=0, le=1)
    other_costs: Decimal = Field(default=Decimal("0"), ge=0)


class SelectionCandidateResponse(ContractModel):
    product_id: str
    title: str
    category_id: str
    category_name: str
    site: SiteCode
    currency: CurrencyCode
    price: Decimal
    cost: Decimal
    shipping_cost: Decimal
    sales_count: int
    rating: Decimal
    review_count: int
    search_index: Decimal | None = None
    sales_index: Decimal | None = None
    competition_index: Decimal | None = None
    growth_rate: Decimal | None = None
    source_name: str
    is_mock_data: bool


class SelectionResultResponse(ContractModel):
    id: UUID
    product_id: str
    title: str | None = None
    rank: int
    total_score: Decimal
    data_completeness: Decimal
    currency: CurrencyCode
    site: SiteCode
    profit: dict[str, object]
    metrics: dict[str, object]
    explanation: dict[str, object]
    risk_warnings: list[str]
    evidence: dict[str, object]
    is_mock_data: bool


class SelectionAnalysisResponse(ContractModel):
    task_id: UUID
    agent_task_id: UUID
    status: AnalysisStatus
    formula_version: str
    generation_mode: Literal["rule_template", "validated_generator"]
    total_candidates: int
    ranked_count: int
    excluded_count: int
    results: list[SelectionResultResponse]
    excluded: list[dict[str, object]]
    is_mock_data: bool


class SelectionTaskResponse(ContractModel):
    task_id: UUID
    agent_task_id: UUID | None
    status: AnalysisStatus
    criteria: dict[str, object]
    formula_version: str
    is_mock_data: bool
    results: list[SelectionResultResponse]


class SelectionExportResponse(ContractModel):
    filename: str
    content_type: Literal["application/json"]
    checksum_sha256: str
    task: SelectionTaskResponse

File: frontend/src/types/selection.ts
export type SiteCode = "sg" | "my" | "ph" | "th" | "vn" | "id";
export type CurrencyCode = "SGD" | "MYR" | "PHP" | "THB" | "VND" | "IDR";
export type SelectionRiskPreference = "conservative" | "balanced" | "growth";
export type SelectionSortField = "sales_count" | "rating" | "review_count" | "price" | "updated_at";

export interface SelectionCandidateQuery {
  site: SiteCode;
  category_id?: string;
  product_ids?: string[];
  min_price?: number;
  max_price?: number;
  sort_by?: SelectionSortField;
  descending?: boolean;
  offset?: number;
  limit?: number;
}

export interface SelectionAnalysisRequest extends SelectionCandidateQuery {
  minimum_profit: number;
  minimum_margin: number;
  platform_fee_rate: number;
  other_costs: number;
  cost_override?: number;
  shipping_cost_override?: number;
  product_weight_kg?: number;
  risk_preference: SelectionRiskPreference;
}

export interface SelectionCandidate {
  product_id: string;
  title: string;
  category_id: string;
  category_name: string;
  site: SiteCode;
  currency: CurrencyCode;
  price: string;
  cost: string;
  shipping_cost: string;
  sales_count: number;
  rating: string;
  review_count: number;
  search_index: string | null;
  sales_index: string | null;
  competition_index: string | null;
  growth_rate: string | null;
  source_name: string;
  is_mock_data: boolean;
}

export interface MetricEvidence {
  score: string | null;
  configured_weight: string;
  effective_weight: string;
  inputs: Record<string, string | number | null>;
  formula: string;
  missing_reason?: string | null;
}

export interface SelectionExplanationEvidence {
  metric: string;
  value: string;
  source: string;
}

export interface SelectionExplanation {
  summary: string;
  evidence: SelectionExplanationEvidence[];
  risks: string[];
  generation_mode: "rule_template" | "validated_generator";
}

export interface SelectionResult {
  id: string;
  product_id: string;
  title: string | null;
  rank: number;
  total_score: string;
  data_completeness: string;
  currency: CurrencyCode;
  site: SiteCode;
  profit: Record<string, string>;
  metrics: Record<string, MetricEvidence>;
  explanation: SelectionExplanation;
  risk_warnings: string[];
  evidence: Record<string, unknown>;
  is_mock_data: boolean;
}

export interface SelectionAnalysis {
  task_id: string;
  agent_task_id: string;
  status: "PENDING" | "RUNNING" | "SUCCEEDED" | "FAILED" | "ARCHIVED";
  formula_version: string;
  generation_mode: "rule_template" | "validated_generator";
  total_candidates: number;
  ranked_count: number;
  excluded_count: number;
  results: SelectionResult[];
  excluded: Record<string, unknown>[];
  is_mock_data: boolean;
}

export interface SelectionTask {
  task_id: string;
  agent_task_id: string | null;
  status: SelectionAnalysis["status"];
  criteria: Record<string, unknown>;
  formula_version: string;
  is_mock_data: boolean;
  results: SelectionResult[];
}

export interface SelectionExport {
  filename: string;
  content_type: "application/json";
  checksum_sha256: string;
  task: SelectionTask;
}
